src/wagtail_ai/ai/base.py

```diff
@@ -0,0 +1,57 @@
+from abc import ABCMeta, abstractmethod
+from collections.abc import Mapping
+from typing import Any, Generic, TypedDict, TypeVar
+
+from ..text_splitters.base import BaseTextSplitterLength
+from ..types import AIResponse, TextSplitterProtocol
+
+
+class ChatModelConfig(TypedDict):
+    id: str
+    token_limit: int
+    extra: Mapping[str, Any] | None
+    text_splitter_class: type[TextSplitterProtocol]
+    text_splitter_length_class: type[BaseTextSplitterLength]
+
+
+AIBackendConfig = TypeVar("AIBackendConfig")
+CMC = TypeVar("CMC", bound=ChatModelConfig)
+
+
+class AIBackend(Generic[AIBackendConfig, CMC], metaclass=ABCMeta):
+    text_splitter_class: type[TextSplitterProtocol]
+    chat_model_config: CMC
+
+    def __init__(
+        self, *, config: AIBackendConfig | None, chat_model_config: CMC
+    ) -> None:
+        self.chat_model_config = chat_model_config
+        self.init_config(backend_config=config, chat_model_config=chat_model_config)
+
+    def init_config(
+        self, *, backend_config: AIBackendConfig | None, chat_model_config: CMC
+    ) -> None:
+        """
+        Take the settings dictionary and parse it onto the backend's instance.
+
+        This method is meant to be overridden by subclasses and is optional.
+        """
+        pass
+
+    @abstractmethod
+    def prompt(self, prompt: str, content: str) -> AIResponse:
```

Comment:
**thought:** I'm not sure about separating 'prompt' and 'content' like this at the `prompt` method level - it suggests that we always need to pass some context along with a prompt and that they are distinct.

In reality there may be cases where the context is sandwiched between two 'prompts', or there is no context at all.

Perhaps we could implement a backend method specific for our rich-text use-cases, something like:

```
def prompt_with_context(self, *, pre_prompt: str, context: str, post_prompt: str) -> AIResponse:
     ...
```

so that if we have a backend that does differentiate system prompts/user prompts/context, we can add specific implementations for those cases?

I'd also suggest we use the term *context* rather than *content* for text that we don't consider the prompt.

Reply:
Yeah, I'm not really clear on the terminology here so I will take your lead on all this stuff.

Whatever happens, this should be an interface that makes sense for Wagtail AI.